File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/types/digitaltwin/ReferenceTargetDescriptor.java
package org.opcfoundation.webserver.types.digitaltwin;

import org.eclipse.milo.opcua.stack.core.types.builtin.LocalizedText;
import org.eclipse.milo.opcua.stack.core.types.builtin.NodeId;
import org.springframework.lang.Nullable;
import org.opcfoundation.webserver.addressspace.nodes.UaObject;
import org.opcfoundation.webserver.digitaltwin.digitaltwin.DigitalTwinType;
import org.opcfoundation.webserver.digitaltwin.element.ElementType;

public class ReferenceTargetDescriptor {
    private final String id;
    private final LocalizedText displayName;
    private final NodeId typeId;
    private @Nullable final UaObject instanceDeclaration;

    public ReferenceTargetDescriptor(
            String digitalTwinId,
            LocalizedText displayName,
            DigitalTwinType digitalTwinType)
    {
        this.id = digitalTwinId;
        this.displayName = displayName;
        this.typeId = digitalTwinType.nodeId();
        this.instanceDeclaration = null;
    }

    public ReferenceTargetDescriptor(
            String elementId,
            LocalizedText displayName,
            ElementType elementType)
    {
        this.id = elementId;
        this.displayName = displayName;
        this.typeId = elementType.nodeId();
        this.instanceDeclaration = null;
    }

    public String getId() {
        return id;
    }

    public LocalizedText getDisplayName() {
        return displayName;
    }

    public NodeId getTypeId() {
        return typeId;
    }

    public @Nullable UaObject getInstanceDeclaration() {
        return instanceDeclaration;
    }
}

File: opcua-webservice-node/source/src/server/types/digitaltwin/ReferenceTargetDescriptor.ts
import { UaLocalizedText, UaNodeId } from "opcua-webapi-ts";
import { UaObject } from "../../addressspace/nodes/UaObject";
import { UaObjectType } from "../../addressspace/nodes/UaObjectType";

export class ReferenceTargetDescriptor {
    private readonly _id: string;
    private readonly _displayName: UaLocalizedText;
    private readonly _typeId: UaNodeId;
    private readonly _instanceDeclaration: UaObject | null;
    
    constructor(
        id: string,
        displayName: UaLocalizedText,
        typeDefinition: UaObjectType,
        instanceDeclaration: UaObject | null = null) {
        this._id = id;
        this._displayName = displayName;
        this._typeId = typeDefinition.nodeId;
        this._instanceDeclaration = instanceDeclaration;
    }

    get id(): string {
        return this._id;
    }

    get displayName(): UaLocalizedText {
        return this._displayName;
    }

    get typeId(): UaNodeId {
        return this._typeId;
    }

    get instanceDeclaration(): UaObject | null {
        return this._instanceDeclaration;
    }
}